x/fdsmr/node.go

```diff
@@ -0,0 +1,121 @@
+// Copyright (C) 2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package fdsmr
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/ava-labs/avalanchego/ids"
+	"github.com/ava-labs/avalanchego/utils/set"
+
+	"github.com/ava-labs/hypersdk/codec"
+	"github.com/ava-labs/hypersdk/x/dsmr"
+)
+
+type Bonder[T dsmr.Tx] interface {
+	// Bond returns if a transaction can be built into a chunk by this node.
+	// If this returns true, Unbond is guaranteed to be called.
+	Bond(tx T) bool
+	// Unbond is called when a tx from an account either expires or is accepted.
+	Unbond(tx T)
+}
+
+// New returns a fortified instance of DSMR
+func New[T dsmr.Interface[U], U dsmr.Tx](d T, b Bonder[U]) *Node[T, U] {
+	return &Node[T, U]{
+		DSMR: d,
+		bonder: bonder[U]{
+			Bonder:  b,
+			pending: set.Set[ids.ID]{},
+		},
+	}
+}
+
+type Node[T dsmr.Interface[U], U dsmr.Tx] struct {
+	DSMR   T
+	bonder bonder[U]
+}
+
+func (n *Node[T, U]) BuildChunk(ctx context.Context, txs []U, expiry int64, beneficiary codec.Address) error {
+	if len(txs) == 0 {
+		return n.DSMR.BuildChunk(ctx, txs, expiry, beneficiary)
+	}
+
```

Comment:
Should we remove this check since it will be handled the same way (slice of size 0 and skip the loop) regardless of if it's here?

Reply:
This is to handle https://github.com/ava-labs/hypersdk/blob/d1e726c6c4416bcce33f71d1b75d67c029a69cd3/x/fdsmr/node_test.go#L36-L45, where we guarantee that `nil` txs are wired into the underlying dsmr as `nil` as well (instead of a zero-length non-nil slice). This is probably overkill but I thought it made sense to be consistent w/ the caller value.